### ISSUE TITLE
Investigate the use of Standardese as a replacement for Doxygen

### ISSUE THREAD

**Issue by ericniebler** [ericniebler/range-v3#460]
https://foonathan.github.io/blog/2016/05/06/standardese-nextgen-doxygen.html

**Comment by foonathan** [comment]
I just noticed it, if you have any feedback/questions feel free to contact me :)

(most likely standardese will have issues parsing it though, but I'll fix any issues that come up)

**Comment by ericniebler** [comment]
@foonathan If you anticipate parsing problems, there's nothing stopping you from preemptively giving it a shot and fixing the bugs. :-)

**Comment by foonathan** [comment]
Guess what I've been doing for the last week :D

Still not quite there yet, but I don't have much time currently sadly.

**Comment by ericniebler** [comment]
😄

**Comment by foonathan** [comment]
I did it, standardese can now successful parse range v3 (albeit uses around 6GB of RAM, it is totally not optimized).
I've put the raw HTML output generated by default on my website, there are two entry points: [by file](https://foonathan.github.io/doc/range-v3/standardese_files.html) or [by (documented) entity](https://foonathan.github.io/doc/range-v3/standardese_entities.html).

I have just took a quick look at your way of documenting (and the error messages about unknown documentation commands issued by standardese).
A quick migration guide:

* `\cond`/`\endcond`: you are using it to hide detail namespaces, the corresponding standardese command is `\exclude` (it will hide the entity and all children, even replacing something like `using foo = detail::bar` with `using foo = implementation-defined`). Note that you can also use the commandline `--input.blacklist_namespace=name` for the same effect globally.

* `\defgroup`, `\ingroup`: grouping is currently not implemented in standardese

* '\sa`/`\relates` etc.: A shorthand for `See XXX` is not implemented, must be manually written out ("See [range::v3::some_type::some_func]()"). `\relates` should not be needed because standardese documents everything declared in the same file together.

* `\return`/`\par`: Those command names are named `\returns` and `\param` by default, you can override any name with the `--comment.cmd_name_XXX=xxx` option though.

If you have any other questions, feel free to ask :)

**Comment by ericniebler** [comment]
> standardese can now successful parse range v3

Awesome! The results look a bit rough, but I see a lot of promise here.

> `\defgroup`, `\ingroup`: grouping is currently not implemented in standardese

I've found grouping and integration with some form of markdown the only way to make automatically-generated documentation usable for non-trivial libraries. I wouldn't call https://ericniebler.github.io/range-v3/ a documentation paragon, but the integration of the hand-written users' guide with the structured and grouped, computer-generated reference section is something worth keeping.

**Comment by foonathan** [comment]
What do you mean by rough besides the weird comments at places?

Grouping will definitely be added (I would have already, but had to fix bugs to support some project...). There will be modules, groups and tags.

Modules are similar to Doxygen groups. Each entity can be in one module and it also applies to all children (so you only need to mark the `view` namespace for example).

Tags are like modules, but they don't apply to children and each entity can have multiple.

Groups are used to group multiple entities together with one implementation, what you are using "@{" for.

That's the plan anyways.

**Comment by ericniebler** [comment]
> What do you mean by rough ... ?

I mean that some interface elements appear to be missing. For example, there is more to the interface of meta::bind_front and bind_back than is suggested here:

https://foonathan.github.io/doc/range-v3/meta__meta.html#meta::v1::bind_back%3CF,%20Us...%3E

And some is completely missing:

https://foonathan.github.io/doc/range-v3/range__v3__view__single.html

Is that because only things with doc comments get emitted in the docs? I'd like to be able to control that.

Will there be a way to integrate the generated reference docs with a markdown processor? Like I said, the ability to generate cross-referenced user docs / reference docs is a very nice feature of Doxygen.

**Comment by foonathan** [comment]
> Is that because only things with doc comments get emitted in the docs? I'd like to be able to control that.

Yes, it should be, option `--input.require_comment=false` to disable.

> Will there be a way to integrate the generated reference docs with a markdown processor?

You can use `--output.format=XXX`, where XXX is `commonmark`, `html`, `latex` or `man`, if that's what you mean.
I'll also need to implement an output template, so that everything is embedded into a file with boilerplate at the top and bottom.

**Comment by foonathan** [comment]
Just a heads up: Turns out the missing entities are a bug, I need to fix that one.

**Comment by foonathan** [comment]
Those bugs are now fixed, so the output is much more complete.

Let's say I'll get back to you when I've implemented more features like grouping?

**Comment by ericniebler** [comment]
Sounds good.

I also not certain the point I made about Markdown integration really landed. Notice how https://ericniebler.github.io/range-v3/ is more than just a reference. It was authored with Markdown (https://github.com/ericniebler/range-v3/blob/master/doc/index.md), and integrated with the generated reference section according to a template I specify (https://github.com/ericniebler/range-v3/blob/master/doc/layout.xml). Doxygen slurps this information in, together with my source files, resolves any cross-references (markdown links -> source, and Doxygen comments in the code -> markdown anchors), and emits the docs as a coherent whole. If your goal is to replace Doxygen, that should be your stretch goal. (I understand that's not a v1 feature.)

**Comment by foonathan** [comment]
Something like this (but more flexible) is planned.

Basically you can write in your markup format and selectively include output generated by standardese. If you use Markdown as your markup format, links will be resolved automatically like they are in the per-entity documentation today.

**Comment by ericniebler** [comment]
Awesome. That's exactly it. Bonus if it can also work in the other direction: from the comments in the source code, I can link to a section in the markdown. That's much less important, though.

**Comment by foonathan** [comment]
Alright it's been implemented in the 0.3 version of standardese: https://foonathan.github.io/blog/2016/11/26/standardese-0.3.html

Linking works in both ways, with the template mode you can easily embed documentation in other files.
